lib/std/crypto/codecs/base64_hex_ct.zig

```diff
@@ -0,0 +1,463 @@
+const StaticBitSet = std.StaticBitSet;
+
+pub const Error = error{
+    /// An invalid character was found in the input.
+    InvalidCharacter,
+    /// The input is not properly padded.
+    InvalidPadding,
+    /// The input buffer is too small to hold the output.
+    NoSpaceLeft,
+    /// The input and output buffers are not the same size.
+    SizeMismatch,
+};
+
+/// (best-effort) constant time hexadecimal encoding and decoding.
+pub const Hex = struct {
+    /// Encodes a binary buffer into a hexadecimal string.
+    /// The output buffer must be twice the size of the input buffer.
+    pub fn encode(hex: []u8, bin: []const u8, comptime case: std.fmt.Case) error{SizeMismatch}!void {
+        if (hex.len / 2 != bin.len) {
+            return error.SizeMismatch;
+        }
+        for (bin, 0..) |v, i| {
+            const b: u16 = v >> 4;
+            const c: u16 = v & 0xf;
+            const off = if (case == .upper) 32 else 0;
+            const x =
+                ((87 - off + c + (((c -% 10) >> 8) & ~@as(u16, 38 - off))) & 0xff) << 8 |
+                ((87 - off + b + (((b -% 10) >> 8) & ~@as(u16, 38 - off))) & 0xff);
+            hex[i * 2] = @truncate(x);
+            hex[i * 2 + 1] = @truncate(x >> 8);
+        }
+    }
+
+    /// Decodes a hexadecimal string into a binary buffer.
+    /// The output buffer must be half the size of the input buffer.
+    pub fn decode(bin: []u8, hex: []const u8) error{ SizeMismatch, InvalidCharacter, InvalidPadding }!void {
+        if (hex.len % 2 != 0) {
+            return error.InvalidPadding;
+        }
+        if (bin.len < hex.len / 2) {
+            return error.SizeMismatch;
+        }
+        _ = decodeAny(bin, hex, null) catch |err| {
+            switch (err) {
+                error.InvalidCharacter => return error.InvalidCharacter,
+                error.InvalidPadding => return error.InvalidPadding,
+                else => unreachable,
+            }
+        };
+    }
+
+    /// A decoder that ignores certain characters.
+    /// The decoder will skip any characters that are in the ignore list.
+    pub const DecoderWithIgnore = struct {
+        /// The characters to ignore.
+        ignored_chars: StaticBitSet(256) = undefined,
+
+        /// Decodes a hexadecimal string into a binary buffer.
+        /// The output buffer must be half the size of the input buffer.
+        pub fn decode(
+            self: DecoderWithIgnore,
+            bin: []u8,
+            hex: []const u8,
+        ) error{ NoSpaceLeft, InvalidCharacter, InvalidPadding }![]const u8 {
+            return decodeAny(bin, hex, self.ignored_chars);
+        }
+
+        /// Returns the decoded length of a hexadecimal string, ignoring any characters in the ignore list.
+        /// This operation doesn't run im constant time, but shouldn't leak information about the actual hexadecimal string.
+        pub fn decodedLenForSlice(Decoder: DecoderWithIgnore, hex: []const u8) !usize {
```

Comment:
this is not a type - https://ziglang.org/documentation/master/#Names

```suggestion
        pub fn decodedLenForSlice(decoder: DecoderWithIgnore, hex: []const u8) !usize {
```

or

```suggestion
        pub fn decodedLenForSlice(self: DecoderWithIgnore, hex: []const u8) !usize {
```